osu.Game/Database/RealmContextFactory.cs

```diff
@@ -113,21 +128,23 @@ protected override void Dispose(bool isDisposing)
         {
             base.Dispose(isDisposing);
 
-            BlockAllOperations();
+            // In the standard case, operations will already be blocked by the Update thread "pausing" from GameHost exit.
+            // This avoids waiting (forever) on an already entered semaphore.
+            if (context != null || active_usages.Value > 0)
```

Comment:
This is the only weird part that I'm not 100% on. This avoids tests potentially sticking on the shutdown phase because of `GameHost` being disposed after it has been stopped.

example:

https://github.com/ppy/osu/blob/b51d0380882c33bae1723af63036533f6ed2d8b3/osu.Game.Tests/Beatmaps/IO/ImportBeatmapTest.cs#L802

Reply:
I don't really follow this case. How is it possible for the semaphore to be "already entered" at this stage? There are only two plausible options - `createContext()` or... another `BlockAllOperations()`?

Reply:
It's the second case. If you look at the test I referenced, it is stopping the game host (first block via thread execution mode change) then disposing the game host - and by extension the game.

This may not be required with the new bindable thread state changes, as I believe we are no longer blocking on `Exiting`, only `Paused`.

Reply:
I tried a framework checkout with the `GameThread` changes and removed this case, and the mentioned test only failed the very first time I ran it ("osu! failed to start in a reasonable amount of time") but then I put it to spin in a loop and it seems fine now, so maybe this is removable indeed...?

Reply:
I've removed this, and also added an explicit exception when disposed (so if we ever accidentally hit such behaviour it doesn't become a debugging nightmare as the thread blocks infinitely).